File: 2024/24-7-2.py
from itertools import product

f = open("24-7.txt").read().split('\n')

target = []
numbers = []

for i in f:
    target.append(int(i.split(':')[0]))
    numbers.append([int(n) for n in i.split(':')[1].split()])

#print(target)
#print(numbers)
   
def can_reach_target_iterative(numbers, target):
    """
    Iterative solution where concatenation applies to the running tally.
    """
    stack = [(0, 0, "")]  # (current_index, current_value, expression)

    while stack:
        index, current_value, expression = stack.pop()

        # If we've used all numbers, check if the target is met
        if index == len(numbers):
            if current_value == target:
                return True, expression
            continue

        # Get the next number
        next_number = numbers[index]

        # Add the next number
        stack.append((index + 1, current_value + next_number, f"{expression} + {next_number}" if expression else str(next_number)))

        # Multiply the next number
        stack.append((index + 1, current_value * next_number if current_value != 0 else next_number, f"{expression} * {next_number}" if expression else str(next_number)))

        # Concatenate the next number with the current_value
        if expression:
            concatenated_value = int(str(current_value) + str(next_number))
            stack.append((index + 1, concatenated_value, f"({expression} || {next_number})"))

    return False, None

tly = 0
for i in range(len(f)):
    found, expression = can_reach_target_iterative(numbers[i], target[i])
    if found:
        print(f"Target {target[i]} can be reached using: {expression}")
    else:
        print(f"Target {target[i]} cannot be reached with the given numbers.")
    if(found):
        tly += target[i]

print('part 2 -',tly)

File: 2024/24-7-2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Diagnostics;

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Enter data in format: target: num1 num2 num3\n...");
        List<string> lines = new List<string>();

        string input;
        while (!string.IsNullOrWhiteSpace(input = Console.ReadLine()))
        {
            lines.Add(input);
        }

        Stopwatch stopwatch = Stopwatch.StartNew(); // Start the timer

        BigInteger total = 0;

        foreach (var line in lines)
        {
            var parts = line.Split(':');
            BigInteger target = BigInteger.Parse(parts[0]);
            List<BigInteger> numbers = parts[1].Split(' ', StringSplitOptions.RemoveEmptyEntries).Select(BigInteger.Parse).ToList();

            var result = CanReachTargetIterative(numbers, target);

            if (result.Found)
            {
                Console.WriteLine($"Target {target} can be reached using: {result.Expression}");
                total += target;
            }
            else
            {
                Console.WriteLine($"Target {target} cannot be reached with the given numbers.");
            }
        }

        stopwatch.Stop(); // Stop the timer
        Console.WriteLine($"Part 2 - Total: {total}");
        Console.WriteLine($"Execution Time: {stopwatch.Elapsed}");
    }

    static (bool Found, string Expression) CanReachTargetIterative(List<BigInteger> numbers, BigInteger target)
    {
        var stack = new Stack<(int Index, BigInteger CurrentValue, string Expression)>();
        stack.Push((0, 0, ""));

        while (stack.Count > 0)
        {
            var (index, currentValue, expression) = stack.Pop();

            if (index == numbers.Count)
            {
                if (currentValue == target)
                {
                    return (true, expression);
                }
                continue;
            }

            BigInteger nextNumber = numbers[index];

            // Add the next number
            stack.Push((index + 1, currentValue + nextNumber, expression == "" ? nextNumber.ToString() : $"{expression} + {nextNumber}"));

            // Multiply the next number
            stack.Push((index + 1, currentValue != 0 ? currentValue * nextNumber : nextNumber, expression == "" ? nextNumber.ToString() : $"{expression} * {nextNumber}"));

            // Concatenate the next number with the current value
            if (!string.IsNullOrEmpty(expression))
            {
                BigInteger concatenatedValue = BigInteger.Parse(currentValue.ToString() + nextNumber.ToString());
                stack.Push((index + 1, concatenatedValue, $"({expression} || {nextNumber})"));
            }
        }

        return (false, null);
    }
}
